### ISSUE TITLE
Bold characters are wider than regular ones when using the Light or SemiLight variants

### ISSUE THREAD

**Issue by CutterXYZ** [microsoft/cascadia-code#713]
### Cascadia family version

2111.01

### Cascadia family variant(s)

Cascadia Code (the version with ligatures), Cascadia Mono (the version without ligatures)

### Font file format(s)

Windows Terminal included version (TTF (variable)), .ttf (variable)

### Platform

Windows 10

### Other Software

Tested in Eclipse IDE (2023-12) and Notepad++ (v8.6).

### What happened?

When using the Light or SemiLight variants, bold text appears wider than regular text, regardless of font size.

Here is a screenshot in Eclipse:
![image](https://github.com/microsoft/cascadia-code/assets/29694177/c46fb795-9f69-4f27-a807-d745acf79224)

In MS Word, there is no difference in width but bold SemiLight text is rendered strangely:
![image](https://github.com/microsoft/cascadia-code/assets/29694177/1e7f6b91-1b9d-4460-ad45-58f0d26dbf60)

**Comment by DHowett** [comment]
Interesting! It looks like Eclipse is unaware that there is a Bold weight for this font, and is _artificially creating a bold typeface_ by stretching out the glyphs.

This is an Eclipse (or perhaps SWT? If they're still using SWT...) bug.

For reference: the Cascadia family has a consistent bounding box size and glyph width for _all_ weights present in the variable font. An application that can correctly map the bold weight to the corresponding named variation will display it just fine.

I suspect that Eclipse's issue is that it can't figure out that "Cascadia Code Bold" is the bold version of "Cascadia Code Light", and is therefore producing a monstrosity we can call "Cascadia Code Light Bold".

**Comment by CutterXYZ** [comment]
This is not just Eclipse. As you can see, Word also does some strange tricks with it. Notepad++ also has the same behavior as Eclipse.

**Comment by aaronbell** [comment]
Unfortunately this is a software limitation and not something (at least as far as I’m aware) that can be solved in the font.

**Comment by DHowett** [comment]
In my response, Eclipse is a representative example. Font family member mapping problems are not unique to Eclipse :smile:

**Comment by CutterXYZ** [comment]
Results seem better with the fixed .ttf versions. Is it recommended to have both versions installed ?

**Comment by DHowett** [comment]
> Results seem better with the fixed .ttf versions. Should I install both versions ?

It might be better to uninstall the variable font, if you can. If not, eh, it should be OK.

---

One of the easier ways to tell whether family-style-axis mapping is working properly is to look at the `$` character¹.

`$` only has a vertical stroke in lighter weights. Therefore, if you see a vertical stroke in your bold font . . . the text editor or UI toolkit that the text editor is using has _invented_ a bold face and ignored the one that comes with the font.

### Working

![image](https://github.com/microsoft/cascadia-code/assets/189190/4861cb6e-36d6-493e-89bc-90736183b62a)

### Broken

![image](https://github.com/microsoft/cascadia-code/assets/189190/909620f0-e244-4400-9207-6dcf296bd9c4)

¹ (Wow, WordPad made it so difficult to generate a Working image that I gave up, Notepad++ wouldn't even load the font (???), and I am not installing Eclipse.)

**Comment by aaronbell** [comment]
It is really quite sad, so many years after the introduction of variable fonts, that app support is still so abysmal.

**Comment by kenmcd** [comment]
Of course it does not work...
There is no style-linking to a heavier weight (as the RIBBI "Bold) in the Light or SemiLight statics, or in the variable fonts.
The Bold weight font is style-linked to the Regular weight font as the "Bold" in that RIBBI style group.
The Light and SemiLight have no actual "Bold" variant linked in their style groups, so the apps are faking it.

We could create another RIBBI style group using the Light (or SemiLight) and the SemiBold (as the "Bold" of the style group) and that should work correctly.
I can post them here if you want to test.

You could also test this with the old Ubuntu fonts (normal.v0.83, not mono) where the Light has the Medium linked as the "Bold". There are two RIBBI style groups in those fonts.

**Comment by CutterXYZ** [comment]
> Of course it does not work... There is no style-linking to a heavier weight (as the RIBBI "Bold) in the Light or SemiLight statics, or in the variable fonts. The Bold weight font is style-linked to the Regular weight font as the "Bold" in that RIBBI style group. The Light and SemiLight have no actual "Bold" variant linked in their style groups, so the apps are faking it.
> 
> We could create another RIBBI style group using the Light (or SemiLight) and the SemiBold (as the "Bold" of the style group) and that should work correctly. I can post them here if you want to test.
> 
> You could also test this with the old Ubuntu fonts (normal.v0.83, not mono) where the Light has the Medium linked as the "Bold". There are two RIBBI style groups in those fonts.

**Comment by CutterXYZ** [comment]
How should I test the Ubuntu font ? It is not monospaced, therefore I don't know if it being wider is normal.

However I would gladly test the modified Cascadia font.

**Comment by CutterXYZ** [comment]
With the static version of the font installed, Word renders the SemiLight bold and regular variants with the same width. This is both with the default static fonts, and the with modified fonts you provided (so, no change).

However Notepad++ renders both versions (default and yours) with a made-up bold variant.

**Comment by CutterXYZ** [comment]
> Unfortunately this is a software limitation and not something (at least as far as I’m aware) that can be solved in the font.

The Regular Bold variant is rendered correctly by all programs. Is there anything different in the relationship between Regular and Regular Bold and the relationship between Semibold and Semilight?

**Comment by aaronbell** [comment]
Yes. There is no relationship between semilight and semibold. They’re just stand-alone instances cut from the variable font rather than designed for use in Office applications which require style-linking.

**Comment by kenmcd** [comment]
> The Regular Bold variant is rendered correctly by all programs. Is there anything different in the relationship between Regular and Regular Bold and the relationship between Semibold and Semilight that kenmcd provided ?

It may be that NotePad++ (by using Scintilla for text rendering) only supports the old RIBZ font family model.
And it may be looking at some "standard" weight and width settings, and it gets confused by anything outside of that old standard 400/700 (such as a Light weight of 300).

If you would like to test this theory...
These are the Light, Light Italic, SemiBold, and SemiBold Italic fonts renamed and configured as an old RIBZ font family.
The weights are set to the "standard" 400/700, and the width is still the standard 5 - but the fonts are still the same visually.
The font base family is now named: CasCodeLight
[CassCodeLight.from.originals.2111.01.(2021-12-14).zip](https://github.com/microsoft/cascadia-code/files/14411913/CassCodeLight.from.originals.2111.01.2021-12-14.zip)
Four fonts:
CasCodeLight-Regular  (R)
CasCodeLight-Italic  (I)
CasCodeLight-Bold  (B)
CasCodeLight-BoldItalic  (Z)

So please try those in NotePad++ and see if they work.

I looked around in the NotePad++ tracker, and the Scintilla tracker and found nothing definitive regarding this, but there are other posts about issues with "non-standard" weights and widths, and some of the limitations of Scintilla.
So this is the only "guess" I can think of which may work.

**Comment by CutterXYZ** [comment]
It works as expected !
![image](https://github.com/microsoft/cascadia-code/assets/29694177/48174171-d71e-4037-82ff-1214b08a3f74)
![image](https://github.com/microsoft/cascadia-code/assets/29694177/5498f2f0-a25b-42ad-a50e-a720ba5050bb)

**Comment by CutterXYZ** [comment]
> > The Regular Bold variant is rendered correctly by all programs. Is there anything different in the relationship between Regular and Regular Bold and the relationship between Semibold and Semilight that kenmcd provided ?
> 
> It may be that NotePad++ (by using Scintilla for text rendering) only supports the old RIBZ font family model. And it may be looking at some "standard" weight and width settings, and it gets confused by anything outside of that old standard 400/700 (such as a Light weight of 300).
> 
> If you would like to test this theory... These are the Light, Light Italic, SemiBold, and SemiBold Italic fonts renamed and configured as an old RIBZ font family. The weights are set to the "standard" 400/700, and the width is still the standard 5 - but the fonts are still the same visually. The font base family is now named: CasCodeLight [CassCodeLight.from.originals.2111.01.(2021-12-14).zip](https://github.com/microsoft/cascadia-code/files/14411913/CassCodeLight.from.originals.2111.01.2021-12-14.zip) Four fonts: CasCodeLight-Regular (R) CasCodeLight-Italic (I) CasCodeLight-Bold (B) CasCodeLight-BoldItalic (Z)
> 
> So please try those in NotePad++ and see if they work.
> 
> I looked around in the NotePad++ tracker, and the Scintilla tracker and found nothing definitive regarding this, but there are other posts about issues with "non-standard" weights and widths, and some of the limitations of Scintilla. So this is the only "guess" I can think of which may work.

Is there a chance of this fix being officially adopted ?

**Comment by kenmcd** [comment]
> Is there a chance of this fix being officially adopted ?

The issue is not in the original fonts (based on these tests).
So the "fix" cannot be officially adopted here.
NotePad++ would need to be fixed to solve this issue.

**Comment by CutterXYZ** [comment]
This "fix" would make the font variants compatible with older software.

**Comment by kenmcd** [comment]
> This "fix" would make the font variants compatible with older software.

Not possible.
NotePad++ expects an old standard RIBBI font family.
These fonts cannot be modified (like I did) to do this without breaking the normal operation of the full family.
The fix needs to come from how NotePad++ handles other styles (weights) used as an alternative to RIBBI.
